### PR TITLE
cannon: Add more sync tests

### LINKED ISSUE #12162: MTCannon: Add more test coverage over Go concurrency primitives
## Description

Add more test coverage over Go concurrency primitives.  In particular, we should make sure to exercise [ll/sc instructions](https://github.com/ethereum-optimism/optimism/pull/11906) as this behavior is new in MTCannon. 

We already have [an existing multithreaded Go test](https://github.com/ethereum-optimism/optimism/blob/829728e095532fbe20779b95f06fada985da54ad/cannon/testdata/example/multithreaded/main.go#L11-L46), but we could increase our test coverage around this.  For example, maybe we have a test that does massively parallel atomic updates.

We should consider porting over the tests from [Go's sync package](https://github.com/golang/go/tree/master/src/sync) wholesale.

### PR DESCRIPTION
**Pull request by mbaxter**
**Description**

Port more [Go sync tests](https://github.com/golang/go/tree/master/src/sync) to increase test coverage over concurrency primitives:
- Add a shared [TestRunner utility](https://github.com/ethereum-optimism/optimism/pull/12949/files#diff-aa6d0ac5e1c01440474fcdacadfa3f95ac61020a351334778124cd4b262e8e1c) that can replace `testing.T` in the Go tests being ported.
- Copy over sync tests from the Go repo (named <orig_file>_copy.go), add a license header, and replace the `testing.T` dependency in each individual tests with the `TestRunner` util.

Builds on: https://github.com/ethereum-optimism/optimism/pull/12771

**Testing**

* Port more unit tests from the Go repo's `sync` package.
* Add some simple tests for the new `TestRunner` utility to make sure we'll get a non-zero exit when a test fails. 

**Metadata**

Fixes https://github.com/ethereum-optimism/optimism/issues/12162

### CODE REVIEW

File: cannon/testdata/utils/testutil/testing.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+func (r *TestRunner) Run(name string, testFunc func(t testing.TB)) bool {
+	testName := r.baseName
+	if name != "" {
+		testName = fmt.Sprintf("%v (%v)", r.baseName, name)
+	}
+
+	var tester testing.TB = r
+	goRunTest(testName, testFunc, tester)
+	return !r.Failed()
```
Comment (Inphi) [inline cannon/testdata/utils/testutil/testing.go:32]: Whenever `r.Fail()` occurs, the current goroutine is terminated. So this doesn't seem reachable on failure. So I want to confirm sure that `r.Fail()` actually results in a non-zero exit code. Is there a caller that needs this specific return value interface?

**Review by Inphi — APPROVED** [review]

File: cannon/testdata/utils/testutil/testing.go (hunk 1 above)
Comment (mbaxter) [reply cannon/testdata/utils/testutil/testing.go:32]: Right, this shouldn't be reachable on failure.  If the test fails, we should [exit the program entirely](https://github.com/ethereum-optimism/optimism/blob/3745516903268e324b6da6ed65f9d8a2c1bec506/cannon/testdata/utils/testutil/testing.go#L48) (fast fail) before we get here with [exitCode == 1](https://github.com/ethereum-optimism/optimism/blob/3745516903268e324b6da6ed65f9d8a2c1bec506/cannon/mipsevm/multithreaded/instrumented_test.go#L68).  The return value shouldn't be needed anywhere - returning the value here just to match the `(*testing.T).Run()` interface.

**Comment by Inphi** [comment]
@mbaxter merging now to lock down audit commit.
